[KBFS-1274] Implement a journaling MDOps

akalin-keybase opened the pull request:
Rename mdserverBranchJournal to mdIDJournal.

Write a mdJournal type on top of mdIDJournal that
handles all the logic with flushing to an MDServer.

Write journalMDOps and have it delegate to a TLF's
mdJournal as needed.

akalin-keybase commented:
OK, at last, this is ready for review.

akalin-keybase commented:
Suggested review order:

1) md_id_journal.go
2) md_journal.go
3) journal_md_ops.go
4) folder_branch_ops.go changes.
5) Everything else.

strib commented on libkbfs/md_id_journal.go:
```diff
@@ -101,11 +100,25 @@ func (j mdServerBranchJournal) readMdID(r MetadataRevision) (MdID, error) {
+	} else if earliestRevision == MetadataRevisionUninitialized {
+		return MdID{}, nil
+	}
+	mdID, err := j.readMdID(earliestRevision)
```
Any reason not to `return j.readMdID(earliestRevision)` as you do below in reading the latest version?

akalin-keybase replied:
Oops, done.

strib commented on libkbfs/md_journal.go:
```diff
@@ -0,0 +1,507 @@
+		return err
+	}
+
+	latestRevision, err := j.j.readLatestRevision()
```
Can't you just use `head.Revision` here?

akalin-keybase replied:
Yeap, done.

strib commented on libkbfs/md_journal.go:
```diff
@@ -0,0 +1,507 @@
+// flushOne sends the earliest MD in the journal to the given MDServer
+// if one exists, and then removes it. Returns whether there was an MD
+// that was put.
+func (j mdJournal) flushOne(
```
I think this function assumes that it will always be called serially from the same goroutine.  Is that right?  Worth a comment, if not a lock.

akalin-keybase replied:
Added a comment for `mdJournal` that it isn't goroutine-safe. (All the other functions rely on this, too.)

strib commented on libkbfs/md_journal.go:
```diff
@@ -0,0 +1,507 @@
+	mStatus := rmd.MergedStatus()
+	bid := rmd.BID
+
+	if (mStatus == Unmerged) && (bid == NullBranchID) &&
```
Maybe a comment here or above the function that it's ok to pass in an unmerged id without the branch ID set, since it gets fixed up here as long as the unmerged flag is set.

strib replied:
Oh actually `rmd.BID` just gets overwritten, it doesn't have to be empty when passed in.  That's a little strange.  Maybe worth a check that we don't override some other bid?

strib replied:
Scratch that.

akalin-keybase replied:
Done.

strib commented on libkbfs/journal_md_ops.go:
```diff
@@ -0,0 +1,382 @@
+	"golang.org/x/net/context"
+)
+
+// A journalMDOps is an implementation of MDOps that delegates to a
```
Get rid of "A ".

akalin-keybase replied:
Done.

strib commented on libkbfs/journal_md_ops.go:
```diff
@@ -0,0 +1,382 @@
+type journalMDOps struct {
+	MDOps
+	jServer  *JournalServer
+	mdServer MDServer
```
As always I'd prefer if this could be changeable dynamically by keeping a `Config` here instead, unless there's a good reason not to.

strib replied:
(I guess you already have access to it in `j.jServer.config`.)

strib replied:
Seems unused actually.

akalin-keybase replied:
Yeap, I think I was going to override `MDServer` at one point, but decided against it. Removed.

strib commented on libkbfs/md_journal.go:
```diff
@@ -0,0 +1,507 @@
+	return head, nil
+}
+
+func (j mdJournal) convertToBranch(
```
It seems possible for a `put()` call to come in halfway through this call, and get written to disk as a merged revision before all of its predecessors are marked as unmerged.  Its `PrevRoot` would also be wrong at that point. I guess when the flush finally gets to that wrong MD, it would fail to put on the server due to server-side checking, but I'm not sure if it would fail due to a revision conflict, or if it could fail due to `PrevRoot` checking or something like that.

Have you thought about that case?  Should we have some locking here so that branch conversion happens more atomically?

akalin-keybase replied:
The locking is done in `journal_md_ops.go`, but I'll double-check that it does what we want here.

akalin-keybase replied:
Yeap, double-checked.

strib commented:
Ok finished reviewing `md_id_journal.go` and `md_journal.go`.  Will resume in the morning with the rest, hopefully.

strib commented on libkbfs/journal_md_ops.go:
```diff
@@ -0,0 +1,382 @@
+	}
+
+	if mStatus == Unmerged && bid != NullBranchID && bid != head.BID {
+		return ImmutableRootMetadata{}, nil
```
Shouldn't it be an error if the branch ID doesn't match the expected branch ID?

akalin-keybase replied:
I had a vague idea that `MDOps` could serve lookups for other branches, but that can't be right. Done.

strib commented on libkbfs/journal_md_ops.go:
```diff
@@ -0,0 +1,382 @@
+
+	rmd := RootMetadata{
+		BareRootMetadata: *head.BareRootMetadata,
+		tlfHandle:        handle,
```
If a non-`nil` `handle` was passed in, should we be checking that it matches the MD?

akalin-keybase replied:
Done. Although eventually I hope to pull `tlfHandle` out of `RootMetadata`.

strib commented on libkbfs/journal_md_ops.go:
```diff
@@ -0,0 +1,382 @@
+
+	lookupTlfID := tlfID
+	if rmd != (ImmutableRootMetadata{}) {
+		lookupTlfID = rmd.ID
```
Error if `rmd.ID != tlfID`?

akalin-keybase replied:
Done.

strib commented on libkbfs/journal_md_ops.go:
```diff
@@ -0,0 +1,382 @@
+}
+
+// TODO: Combine the two GetForTLF functions in MDOps to avoid the
+// need for these helper functions.
```
I only see one helper?

strib replied:
Ah, guess you're referring to `getRange`.

akalin-keybase replied:
Changed comment, since `getRange` has its own helper.

strib commented on libkbfs/journal_md_ops.go:
```diff
@@ -0,0 +1,382 @@
+
+	// Otherwise, fetch the rest from the server and prepend them.
+	serverStop := jirmds[0].Revision - 1
+	irmds, err := delegateFn(ctx, id, start, stop)
```
`stop` -> `serverStop`?  If a flush happens, or if another client pushed a revision, you could hit the error case below I think.  Maybe add a test to cover this case?

akalin-keybase replied:
D'oh, nice catch. I'm punting on more tests for `journalMdOps` (and to clean up the existing one) for a future PR, but added a TODO.

strib commented on libkbfs/journal_md_ops_test.go:
```diff
@@ -0,0 +1,209 @@
+	}()
+
+	config := MakeTestConfigOrBust(t, "test_user")
+	defer config.Shutdown()
```
Why not `CheckConfigAndShutdown()`?  Are there expected errors?   If so, a comment explaining would be good.

akalin-keybase replied:
No reason, done.

strib commented on libkbfs/folder_branch_ops.go:
```diff
@@ -1849,7 +1847,7 @@ func (fbo *folderBranchOps) finalizeBlocks(bps *blockPutState) error {
 
 // Returns true if the passed error indicates a revision conflict.
-func (fbo *folderBranchOps) isRevisionConflict(err error) bool {
+func isRevisionConflict(err error) bool {
```
I think this will conflict with one of my outstanding PRs, just FYI.  May the best man win.

strib commented on libkbfs/interfaces.go:
```diff
@@ -782,14 +786,10 @@ type MDOps interface {
 	// user has permission to do so.
-	GetForHandle(ctx context.Context, handle *TlfHandle) (
+	GetForHandle(
+		ctx context.Context, handle *TlfHandle, mStatus MergeStatus) (
```
Mention `mStatus` in function comment.

akalin-keybase replied:
Done.

strib commented on libkbfs/md_util.go:
```diff
@@ -219,6 +221,84 @@ func getUnmergedMDUpdates(ctx context.Context, config Config, id TlfID,
+func encryptMDPrivateData(
+	ctx context.Context, codec Codec, crypto cryptoPure,
+	signer cryptoSigner, ekg encryptionKeyGetter, me keybase1.UID,
+	rmd ReadOnlyRootMetadata, brmd *BareRootMetadata) error {
```
`brmd` is basically a return parameter right?  Any reason not to make it that?

akalin-keybase replied:
I wanted to save a copy when we're building a `RMDSigned`, but that's not really worth it. Changed.

strib commented on libkbfs/kbfs_ops_test.go:
```diff
@@ -995,23 +1017,14 @@ func expectSyncBlockHelper(
+		oldMDOps := config.MDOps()
+		if oldShim, ok := oldMDOps.(shimMDOps); ok {
+			if oldShim.isUnmerged != isUnmerged {
+				panic("old shim with different isUnmerged")
```
Weird to panic in a test.  Why not `t.Fatal` so we get the preceding log dump?

strib replied:
Actually I guess `panic` gives you the log dump too.  But I still think a fatal would be better.

akalin-keybase replied:
Changed to `t.Fatal`, also changed a few other panics.

strib commented:
Cool!  Thanks for getting this beast together.  Bunch of questions inline, but mostly looks good.  High-level things:

* When journaling gets turned on by default, it looks to me like we'll no longer have any truly diskless mode for `-server-in-memory`.  Is that right?  Might want to turn off journaling when that flag is passed in (though probably doesn't need to change for this PR).
* I'd like a little more logging in the journal, especially when a new revision is put and when an old one is flushed.
* If we ever hit a bad/slow disk, we will be unresponsive to ctrl-C.  What do you think about wrapping the file IO in something that takes a context and forks a goroutine for the actual disk-blocking stuff?  Would that be too much?

akalin-keybase commented on libkbfs/journal_md_ops.go:
```diff
@@ -0,0 +1,382 @@
+			return err
+		}
+		if irmd.BID == bid {
+			err := bundle.mdJournal.clear()
```
Fixed, by doing the head lookup in `clear`.

akalin-keybase commented:
- Added a TODO for `-server-in-memory`.
- Added more logging in `mdJournal`.
- I had a TODO to make disk I/O cancellable somewhere, but I also filed a story for it. I agree it should be done, but it can wait, I think.

akalin-keybase commented:
Okay, addressed all comments. Probably worth another quick look from you before you leave!

strib commented on libkbfs/journal_md_ops.go:
```diff
@@ -376,16 +395,20 @@ func (j journalMDOps) PruneBranch(
+			bundle.lock.Lock()
+			defer bundle.lock.Unlock()
+
+			_, uid, err := j.jServer.config.KBPKI().GetCurrentUserInfo(ctx)
```
No need to hold the lock over this RPC.

akalin-keybase replied:
Done.

strib commented on libkbfs/md_journal.go:
```diff
@@ -506,6 +509,16 @@ func (j mdJournal) flushOne(
 }
 
-func (j mdJournal) clear() error {
+func (j mdJournal) clear(currentUID keybase1.UID, bid BranchID) error {
```
Might as well log in here too.

akalin-keybase replied:
Done.

strib commented on libkbfs/md_journal.go:
```diff
@@ -0,0 +1,524 @@
+// modifying it as needed. In particular, if this is an unmerged
+// RootMetadata but the branch ID isn't set, it will be set to the
+// journal's branch ID, which is assumed to be non-zero.
+func (j mdJournal) put(
```
Add a log message in here somewhere?

akalin-keybase replied:
Somehow I forgot to push a last commit which actually added the logs. Done.

strib commented:
Looks good to me mod log nits, thanks!

akalin-keybase commented:
CI failure looks unrelated.
